### PR TITLE
Add REST API Blog Auth & Stats function using it.

### PR DESCRIPTION
**Pull request by georgestephanis**
Major props to @jblz who had done the grunt of this lifting.

To Test:

On WPCOM push through a commit to the `$this->jetpack_auth_allowed_sites` on or about line 87 of `public.api/rest/class.wpcom-json-api.php` adding the site id that you will be testing with.

Then, from the Jetpack site whose id was whitelisted for testing, run this from the command line:

```
wp eval "print_r( stats_get_from_restapi() );"
```

through wp-cli.  It should give you good data back.

This pull request does NOT in any way account for a gui, it is merely providing utility functions.

### CODE REVIEW

File: class.jetpack-client.php [hunk 1]
```diff
@@ -216,4 +219,36 @@ public static function set_time_diff( &$response, $force_set = false ) {
...
+	 * Query the WordPress.com REST API using the blog token
+	 */
+	static function wpcom_json_api_request_as_blog( $path, $version = self::WPCOM_JSON_API_VERSION, $args = array(), $body = null ) {
+		$filtered_args = array_intersect_key( $args, array(
+			'method'      => 'string',
+			'timeout'     => 'int',
+			'redirection' => 'int',
+		) );
+
+		$proto = apply_filters( 'jetpack_can_make_outbound_https', true ) ? 'https' : 'http';
```
Comment (jeherve) [inline class.jetpack-client.php]: We'll need to add a docblock here.

Comment (georgestephanis) [inline class.jetpack-client.php]: Drat, knew I'd missed something.

Comment (georgestephanis) [inline class.jetpack-client.php]: Actually -- @jblz -- was  this filter meant to tie into something specifically?  I'm not seeing that filter defined anywhere currently.

**Comment by georgestephanis** [comment]
Added phpdoc to filter, but I'm still a bit confused as to the point of the new `jetpack_can_make_outbound_https` filter.

**Comment by dereksmart** [comment]
This is excellent.  Tested, tried and true.

File: modules/stats.php [hunk 2]
```diff
@@ -1226,3 +1226,71 @@ function stats_str_getcsv( $csv ) {
...
+/**
+ * Fetches stats data from the REST API.  Caches locally for 5 minutes.
+ *
+ * @link: https://developer.wordpress.com/docs/api/1.1/get/sites/%24site/stats/
+ *
+ * @param  array|string   $args     The args that are passed to the endpoint
+ * @param  string         $resource Optional sub-endpoint following /stats/
+ * @return array|WP_Error
+ */
+function stats_get_from_restapi( $args = array(), $resource = '' ) {
```
Comment (jblz) [inline modules/stats.php]: should this function name be prefixed by `jetpack_`?

File: class.jetpack-client.php (hunk 1 above)
Comment (jblz) [inline class.jetpack-client.php]: @georgestephanis Honestly, I can't remember what `jetpack_can_make_outbound_https` was supposed to do :)

I _think_ I remember someone mentioning (on the API side of the discussion) that not all hosts are capable of outbound https and maybe this was some half formed solution..?

**Comment by jblz** [comment]
Tests well here, as well. Nice work.
:+1:

File: modules/stats.php (hunk 2 above)
Comment (dereksmart) [inline modules/stats.php]: cc @georgestephanis

Comment (georgestephanis) [inline modules/stats.php]: I don't think so -- it's prefixed by stats, which matches the legacy `stats_get_csv()` call

**Pull request merged**